File: LeetCode/560.Subarray_Sum_Equals_K/560.Subarray_Sum_Equals_K.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        // prefixsum + map
        // map key: prefixSum, value: count
        // (currentSum - k) == key,if exist, result += count
        int currentSum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int result = 0;
        for (int i = 0; i < nums.length; i++) {
            currentSum += nums[i];
            int diff = currentSum - k;
            if (map.containsKey(diff)) {
                result += map.get(diff);
            }
            int count = map.getOrDefault(currentSum, 0) + 1;
            map.put(currentSum, count);
        }

        return result;
    }
}

File: LeetCode/560.Subarray_Sum_Equals_K/560.Subarray_Sum_Equals_K.swift
class Solution {
    func subarraySum(_ nums: [Int], _ k: Int) -> Int {
        var last = 0
        var dic = [0: 1]

        var res = 0
        for num in nums {
            last += num
            if let cnt = dic[last - k] {
                res += cnt
            }
            dic[last, default: 0] += 1
        }

        return res
    }
}
